lib/services/resource.go

```diff
@@ -714,6 +720,67 @@ func UnmarshalResource(kind string, raw []byte, opts ...MarshalOption) (types.Re
 	return u, nil
 }
 
+// ProtoResource abstracts a resource defined as a protobuf message.
+type ProtoResource interface {
+	proto.Message
+	// GetMetadata returns the generic resource metadata.
+	GetMetadata() *headerv1.Metadata
+}
+
+// MarshalProtoResource marshals a ProtoResource to JSON.
+func MarshalProtoResource[T ProtoResource](resource T, opts ...MarshalOption) ([]byte, error) {
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if !cfg.PreserveResourceID {
+		resource = proto.Clone(resource).(T)
+		//nolint:staticcheck // SA1019. Id is deprecated, but still needed.
+		resource.GetMetadata().Id = 0
+		resource.GetMetadata().Revision = ""
+	}
+	data, err := utils.FastMarshal(resource)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return data, nil
+}
+
+// ProtoResourcePtr is a ProtoResource that is also a pointer to T.
+type ProtoResourcePtr[T any] interface {
+	*T
+	ProtoResource
+}
+
+// UnmarshalProtoResource unmarshals a ProtoResource from JSON. It is paramaterized on types T and U, where T
+// is a pointer type that implements ProtoResource, and U is the type that T points to. This is so that it can
+// allocate and instance of U to unmarshal into, without reflection.
+func UnmarshalProtoResource[T ProtoResourcePtr[U], U any](data []byte, opts ...MarshalOption) (T, error) {
+	if len(data) == 0 {
+		return nil, trace.BadParameter("nothing to unmarshal")
+	}
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var resource T = new(U)
+	err = utils.FastUnmarshal(data, resource)
```

Comment:
should you use protojson since it's a protobuf message?

Reply:
We should keep others with the old function but new ones should use protojson

@rosstimothy what's your take?

Reply:
Alright, i think it was still good to refactor these 7 identical functions into one, so I kept that and marked it deprecated, and added new protojson-based marshaler/unmarshaler